File: apps/android/app/src/main/kotlin/sh/nikhil/swekitty/AppearanceStore.kt
package sh.nikhil.swekitty

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * User-tunable appearance settings: chat body font, theme override,
 * and turn-collapse preference. Persisted to plain SharedPreferences
 * (these aren't secrets — no encryption needed).
 *
 * Mirrors `apps/ios/Sources/Models/AppearanceStore.swift`. Plan called
 * for DataStore but SharedPreferences keeps the dependency surface
 * smaller and is consistent with [SessionStore]'s prefs pattern.
 */
class AppearanceStore : ViewModel() {

    enum class FontFamily(val label: String) {
        Monospaced("Monospaced"),
        System("System"),
    }

    enum class ThemeMode(val label: String) {
        System("System"),
        Light("Light"),
        Dark("Dark"),
    }

    /**
     * Palette choice for the "Neon Terminal" theme system. Enum-name
     * persistence is by the stable [id] string (not [name]) so it
     * matches iOS `NeonPaletteChoice` / `NeonPalette` and the Compose
     * `NeonPalette.id` one-for-one. Tokens are resolved by
     * `sh.nikhil.swekitty.ui.NeonTheme.resolve(...)`; the effective
     * dark/light is reused from [themeMode] (no separate neon mode).
     */
    enum class NeonPalette(val id: String, val label: String) {
        Ice("ice", "Ice"),
        Synth("synth", "Synthwave"),
        Matrix("matrix", "Matrix"),
        Amber("amber", "Amber CRT");

        companion object {
            fun fromId(id: String?): NeonPalette =
                entries.firstOrNull { it.id == id } ?: Ice
        }
    }

    /**
     * Color theme for the terminal renderer. Mirrors iOS
     * `GhosttyVT.GhosttyTheme` / `AppearanceStore.GhosttyTerminalTheme`
     * one-for-one — same five curated themes, and the concrete
     * `#rrggbb` values live in [sh.nikhil.swekitty.ui.TerminalPalette]
     * (read verbatim from the iOS source) so both platforms render
     * identically. Applied to the xterm.js path ([WebTerminal]) and the
     * Termux path ([TermuxTerminalView]) alike. Persisted by enum name.
     */
    enum class TerminalTheme(val label: String) {
        GhosttyDark("Ghostty Dark"),
        SolarizedDark("Solarized Dark"),
        Nord("Nord"),
        Dracula("Dracula"),
        GruvboxDark("Gruvbox Dark"),
    }

    private val _fontFamily = MutableStateFlow(FontFamily.Monospaced)
    val fontFamily: StateFlow<FontFamily> = _fontFamily.asStateFlow()

    private val _themeMode = MutableStateFlow(ThemeMode.System)
    val themeMode: StateFlow<ThemeMode> = _themeMode.asStateFlow()

    private val _collapseTurns = MutableStateFlow(false)
    val collapseTurns: StateFlow<Boolean> = _collapseTurns.asStateFlow()

    /**
     * Stage 0 feature flag for the Termux `terminal-view` native
     * terminal path. Mirrors iOS `experimentalNativeTerminal`. Off by
     * default — the xterm.js path ([WebTerminal]) remains the
     * production renderer until Stage 2 of the rewrite ships. See
     * `docs/PLAN-TERMINAL-REWRITE.md` (Android section).
     */
    private val _experimentalNativeTerminal = MutableStateFlow(false)
    val experimentalNativeTerminal: StateFlow<Boolean> = _experimentalNativeTerminal.asStateFlow()

    /**
     * Body point size for the chat typography ramp (Android mirror of
     * iOS [AppearanceStore.bodyPointSize], landed alongside the
     * Settings → Font Size slider in PLAN-LITTER-VISUAL-PARITY PR 2).
     * Range is [BODY_POINT_SIZE_RANGE]; setters clamp out-of-range
     * writes so corrupted prefs cannot blow out the layout.
     */
    private val _bodyPointSize = MutableStateFlow(DEFAULT_BODY_POINT_SIZE)
    val bodyPointSize: StateFlow<Float> = _bodyPointSize.asStateFlow()

    /**
     * Terminal cell font size in points. Android mirror of iOS
     * [AppearanceStore.ghosttyFontSize]. Default is a dense
     * [DEFAULT_TERMINAL_FONT_SIZE] (10pt) so a real-terminal grid fits
     * on a phone, matching iOS; range is [TERMINAL_FONT_SIZE_RANGE].
     * Drives the xterm.js `fontSize` option (re-fit on change) and the
     * Termux cell text size. Setters clamp out-of-range writes.
     */
    private val _terminalFontSize = MutableStateFlow(DEFAULT_TERMINAL_FONT_SIZE)
    val terminalFontSize: StateFlow<Float> = _terminalFontSize.asStateFlow()

    /**
     * Curated terminal color theme. Default [TerminalTheme.GhosttyDark]
     * matches iOS. Drives the xterm.js `theme` option and the Termux
     * colour table. Persisted by enum name.
     */
    private val _terminalTheme = MutableStateFlow(TerminalTheme.GhosttyDark)
    val terminalTheme: StateFlow<TerminalTheme> = _terminalTheme.asStateFlow()

    /**
     * Neon Terminal palette choice. Default [NeonPalette.Ice], matching
     * iOS. Persisted by [NeonPalette.id]; resolved into a [NeonTheme]
     * and provided via `LocalNeonTheme` in MainActivity.
     */
    private val _neonPalette = MutableStateFlow(NeonPalette.Ice)
    val neonPalette: StateFlow<NeonPalette> = _neonPalette.asStateFlow()

    /**
     * Neon Terminal glow on/off. Default `true`, matching iOS. Flows
     * into `NeonTheme.resolve(...)` so later card work can render (or
     * skip) the layered glow shadows.
     */
    private val _neonGlow = MutableStateFlow(true)
    val neonGlow: StateFlow<Boolean> = _neonGlow.asStateFlow()

    private var prefs: SharedPreferences? = null

    fun hydrate(ctx: Context) {
        val p = ctx.getSharedPreferences("swekitty.appearance", Context.MODE_PRIVATE)
        prefs = p
        _fontFamily.value = p.getString(KEY_FONT, null)
            ?.let { runCatching { FontFamily.valueOf(it) }.getOrNull() }
            ?: FontFamily.Monospaced
        _themeMode.value = p.getString(KEY_THEME, null)
            ?.let { runCatching { ThemeMode.valueOf(it) }.getOrNull() }
            ?: ThemeMode.System
        _collapseTurns.value = p.getBoolean(KEY_COLLAPSE, false)
        _experimentalNativeTerminal.value = p.getBoolean(KEY_EXPERIMENTAL_NATIVE_TERMINAL, false)
        _bodyPointSize.value = p.getFloat(KEY_BODY_POINT_SIZE, DEFAULT_BODY_POINT_SIZE)
            .coerceIn(BODY_POINT_SIZE_RANGE)
        _terminalFontSize.value = p.getFloat(KEY_TERMINAL_FONT_SIZE, DEFAULT_TERMINAL_FONT_SIZE)
            .coerceIn(TERMINAL_FONT_SIZE_RANGE)
        _terminalTheme.value = p.getString(KEY_TERMINAL_THEME, null)
            ?.let { runCatching { TerminalTheme.valueOf(it) }.getOrNull() }
            ?: TerminalTheme.GhosttyDark
        _neonPalette.value = NeonPalette.fromId(p.getString(KEY_NEON_PALETTE, null))
        _neonGlow.value = p.getBoolean(KEY_NEON_GLOW, true)
    }

    fun setFontFamily(value: FontFamily) {
        _fontFamily.value = value
        prefs?.edit()?.putString(KEY_FONT, value.name)?.apply()
    }

    fun setThemeMode(value: ThemeMode) {
        _themeMode.value = value
        prefs?.edit()?.putString(KEY_THEME, value.name)?.apply()
    }

    fun setCollapseTurns(value: Boolean) {
        _collapseTurns.value = value
        prefs?.edit()?.putBoolean(KEY_COLLAPSE, value)?.apply()
    }

    fun setExperimentalNativeTerminal(value: Boolean) {
        _experimentalNativeTerminal.value = value
        prefs?.edit()?.putBoolean(KEY_EXPERIMENTAL_NATIVE_TERMINAL, value)?.apply()
    }

    /**
     * Set body point size, clamped into [BODY_POINT_SIZE_RANGE]. Mirrors
     * the iOS setter: silent clamp on out-of-range writes so a slider
     * with rounding error or a corrupted pref cannot blow out the
     * layout.
     */
    fun setBodyPointSize(value: Float) {
        val clamped = value.coerceIn(BODY_POINT_SIZE_RANGE)
        _bodyPointSize.value = clamped
        prefs?.edit()?.putFloat(KEY_BODY_POINT_SIZE, clamped)?.apply()
    }

    /** Set terminal font size, clamped into [TERMINAL_FONT_SIZE_RANGE]
     *  (mirrors the iOS [ghosttyFontSize] setter's silent clamp). */
    fun setTerminalFontSize(value: Float) {
        val clamped = value.coerceIn(TERMINAL_FONT_SIZE_RANGE)
        _terminalFontSize.value = clamped
        prefs?.edit()?.putFloat(KEY_TERMINAL_FONT_SIZE, clamped)?.apply()
    }

    fun setTerminalTheme(value: TerminalTheme) {
        _terminalTheme.value = value
        prefs?.edit()?.putString(KEY_TERMINAL_THEME, value.name)?.apply()
    }

    /** Set the Neon Terminal palette; persisted by stable [NeonPalette.id]. */
    fun setNeonPalette(value: NeonPalette) {
        _neonPalette.value = value
        prefs?.edit()?.putString(KEY_NEON_PALETTE, value.id)?.apply()
    }

    /** Set the Neon Terminal glow flag. */
    fun setNeonGlow(value: Boolean) {
        _neonGlow.value = value
        prefs?.edit()?.putBoolean(KEY_NEON_GLOW, value)?.apply()
    }

    companion object {
        /** Clamp range for [bodyPointSize] (matches iOS). */
        val BODY_POINT_SIZE_RANGE: ClosedFloatingPointRange<Float> = 12f..18f
        /** Default body point size on a fresh install (matches iOS). */
        const val DEFAULT_BODY_POINT_SIZE: Float = 14f

        /** Clamp range for [terminalFontSize] (matches iOS
         *  `ghosttyFontSizeRange` 8...24). */
        val TERMINAL_FONT_SIZE_RANGE: ClosedFloatingPointRange<Float> = 8f..24f
        /** Default terminal font size — a dense 10pt, matching iOS
         *  `defaultGhosttyFontSize`. Denser than the old 13pt xterm.js
         *  default so a real-terminal grid fits on a phone. */
        const val DEFAULT_TERMINAL_FONT_SIZE: Float = 10f

        // SharedPreferences keys — kept private (file-scope) so callers
        // go through the typed setters / state flows above. Live in the
        // public companion so we can have just one (Kotlin only allows a
        // single companion object per class).
        private const val KEY_FONT = "font"
        private const val KEY_THEME = "theme"
        private const val KEY_COLLAPSE = "collapseTurns"
        private const val KEY_EXPERIMENTAL_NATIVE_TERMINAL = "experimentalNativeTerminal"
        private const val KEY_BODY_POINT_SIZE = "bodyPointSize"
        private const val KEY_TERMINAL_FONT_SIZE = "terminalFontSize"
        private const val KEY_TERMINAL_THEME = "terminalTheme"
        private const val KEY_NEON_PALETTE = "neonPalette"
        private const val KEY_NEON_GLOW = "neonGlow"
    }
}

/**
 * CompositionLocal so any composable below `AppRoot` can read
 * appearance without threading the store through every parameter list.
 */
val LocalAppearanceStore = staticCompositionLocalOf<AppearanceStore> {
    error("AppearanceStore not provided")
}

File: apps/ios/Sources/Models/AppearanceStore.swift
import Foundation
import Observation
import SwiftUI
import UIKit

/// User-tunable appearance settings: chat body font, theme override,
/// and turn-collapse preference. Persisted to `UserDefaults.standard`
/// so the choice survives relaunch.
///
/// Lives at app root, injected as an `@Environment` value into any
/// view that needs to honour it (currently `ConversationView` for the
/// monospaced body font, `SettingsSheet`/`AppearanceSheet` for the UI).
@Observable
final class AppearanceStore {
    enum FontFamily: String, CaseIterable, Identifiable {
        case serif
        case system
        case monospaced

        var id: String { rawValue }
        var label: String {
            switch self {
            case .serif:      return "Serif"
            case .system:     return "System"
            case .monospaced: return "Monospaced"
            }
        }
    }

    /// Color theme for the experimental native (libghostty) terminal.
    /// Self-contained mirror of `GhosttyVT.GhosttyTheme` — same rawValues
    /// so `GhosttyTerminalView` can map across the module boundary with a
    /// plain `init(rawValue:)`. Kept here (rather than re-exporting the
    /// GhosttyVT enum) so the model layer + its tests don't have to link
    /// libghostty. Only applies on the `experimentalNativeTerminal` path.
    enum GhosttyTerminalTheme: String, CaseIterable, Identifiable {
        case ghosttyDark
        case solarizedDark
        case nord
        case dracula
        case gruvboxDark

        var id: String { rawValue }
        var label: String {
            switch self {
            case .ghosttyDark:   return "Ghostty Dark"
            case .solarizedDark: return "Solarized Dark"
            case .nord:          return "Nord"
            case .dracula:       return "Dracula"
            case .gruvboxDark:   return "Gruvbox Dark"
            }
        }
    }

    /// Palette choice for the "Neon Terminal" theme system. RawValues
    /// are the stable persistence ids and match `NeonPalette` /
    /// Android `NeonPalette.id` one-for-one. The resolved tokens live in
    /// `NeonTheme.resolve(...)`; the effective dark/light comes from
    /// `themeMode` (reused — there is no separate neon mode setting).
    enum NeonPaletteChoice: String, CaseIterable, Identifiable {
        case ice
        case synth
        case matrix
        case amber

        var id: String { rawValue }
        var label: String {
            switch self {
            case .ice:    return "Ice"
            case .synth:  return "Synthwave"
            case .matrix: return "Matrix"
            case .amber:  return "Amber CRT"
            }
        }

        /// Bridge to the resolved-token enum in `NeonTheme.swift`. Kept
        /// as a 1:1 rawValue mapping so the model layer (+ its tests)
        /// doesn't have to depend on the Theme layer's type.
        var neonPalette: NeonPalette { NeonPalette(rawValue: rawValue) ?? .ice }
    }

    enum ThemeMode: String, CaseIterable, Identifiable {
        case system
        case light
        case dark

        var id: String { rawValue }
        var label: String {
            switch self {
            case .system: return "System"
            case .light:  return "Light"
            case .dark:   return "Dark"
            }
        }

        var colorScheme: ColorScheme? {
            switch self {
            case .system: return nil
            case .light:  return .light
            case .dark:   return .dark
            }
        }
    }

    private enum Keys {
        static let font = "conduit.appearance.font"
        static let theme = "conduit.appearance.theme"
        static let collapseTurns = "conduit.appearance.collapseTurns"
        /// Stage 0 feature flag for the Ghostty-libghostty native
        /// terminal path. See docs/PLAN-TERMINAL-REWRITE.md. Defaults
        /// off; xterm.js stays the production renderer until Stage 2.
        static let experimentalNativeTerminal = "conduit.experimental.nativeTerminal"
        /// Trash-rebuild feature flag for the parallel `ConduitUI/` view
        /// tree. When on, `ConduitApp` renders `ConduitUI.RootView`
        /// instead of the current `RootView`. Off by default for this
        /// PR — follow-up PRs flip the default and delete the old
        /// views. See `docs/PLAN-CONDUIT-UI.md`.
        static let experimentalConduitUI = "conduit.experimental.conduitUI"
        /// Body point size for the typography ramp
        /// (`ConduitTypography`). User-tunable within
        /// [bodyPointSizeRange]; everything in the ramp scales off this.
        static let bodyPointSize = "conduit.appearance.bodyPointSize"
        /// Font size (points) libghostty renders the native terminal grid
        /// at. Only consumed on the `experimentalNativeTerminal` path.
        static let ghosttyFontSize = "conduit.appearance.ghosttyFontSize"
        /// Color theme rawValue for the native (libghostty) terminal.
        /// Only consumed on the `experimentalNativeTerminal` path.
        static let ghosttyTerminalTheme = "conduit.appearance.ghosttyTerminalTheme"
        /// Palette choice for the Neon Terminal theme system
        /// (`NeonPaletteChoice` rawValue). Resolved into tokens by
        /// `NeonTheme.resolve(...)` and injected via `\.neonTheme`.
        static let neonPalette = "conduit.appearance.neonPalette"
        /// Glow on/off toggle for the Neon Terminal theme system.
        static let neonGlow = "conduit.appearance.neonGlow"
    }

    /// Clamp range for the native-terminal font size. Lower bound keeps a
    /// dense grid legible on iPhone; upper bound stops the grid collapsing
    /// to a few columns. Stepper steps by 1.
    static let ghosttyFontSizeRange: ClosedRange<Double> = 8...24
    /// Default native-terminal font size. The 13pt default tested too
    /// large on device; 10pt gives a denser, real-terminal feel while
    /// staying legible on iPhone. Users can still adjust via the 8–24
    /// Settings slider.
    static let defaultGhosttyFontSize: Double = 10

    /// Clamp range for [bodyPointSize]. Lower bound keeps captions
    /// readable; upper bound prevents headings from blowing out the
    /// composer / list rows.
    static let bodyPointSizeRange: ClosedRange<CGFloat> = 12...18
    /// Default chosen to match upstream's `ConduitFont.conversationBodyPointSize`
    /// starting value at the centre of the slider's range.
    static let defaultBodyPointSize: CGFloat = 14

    var fontFamily: FontFamily {
        didSet { defaults.set(fontFamily.rawValue, forKey: Keys.font) }
    }

    var themeMode: ThemeMode {
        didSet {
            defaults.set(themeMode.rawValue, forKey: Keys.theme)
            applyToWindows()
        }
    }

    var collapseTurns: Bool {
        didSet { defaults.set(collapseTurns, forKey: Keys.collapseTurns) }
    }

    /// Stage 0 feature flag — when on, the Terminal tab renders via the
    /// experimental Ghostty-libghostty path (`GhosttyTerminalView`)
    /// instead of the production xterm.js path (`TerminalTabXterm`).
    /// See `docs/PLAN-TERMINAL-REWRITE.md`. The xterm.js path stays
    /// the default until Stage 2 of that plan ships.
    var experimentalNativeTerminal: Bool {
        didSet { defaults.set(experimentalNativeTerminal, forKey: Keys.experimentalNativeTerminal) }
    }

    /// Trash-rebuild flag — when true, the app boots into the parallel
    /// `ConduitUI` view tree rather than the legacy `RootView`. Default
    /// `false`; users opt in via Settings → Experimental → "Conduit UI
    /// (preview)". See `apps/ios/Sources/ConduitUI/` and
    /// `docs/PLAN-CONDUIT-UI.md`.
    var experimentalConduitUI: Bool {
        didSet { defaults.set(experimentalConduitUI, forKey: Keys.experimentalConduitUI) }
    }

    /// Font size libghostty renders the native terminal at. Setter
    /// clamps into [ghosttyFontSizeRange] and persists. A change here is
    /// picked up live by `GhosttyTerminalView` (config update + PTY-grid
    /// resync). Only applies on the `experimentalNativeTerminal` path.
    var ghosttyFontSize: Double = AppearanceStore.defaultGhosttyFontSize {
        didSet {
            let clamped = ghosttyFontSize.clamped(to: Self.ghosttyFontSizeRange)
            if clamped != ghosttyFontSize {
                ghosttyFontSize = clamped
                return
            }
            defaults.set(ghosttyFontSize, forKey: Keys.ghosttyFontSize)
        }
    }

    /// Color theme for the native (libghostty) terminal. Persisted by
    /// rawValue; applied live by `GhosttyTerminalView`. Only applies on
    /// the `experimentalNativeTerminal` path.
    var ghosttyTerminalTheme: GhosttyTerminalTheme {
        didSet { defaults.set(ghosttyTerminalTheme.rawValue, forKey: Keys.ghosttyTerminalTheme) }
    }

    /// Neon Terminal palette choice. Persisted by rawValue; resolved
    /// into a `NeonTheme` at the app root and injected via the
    /// `\.neonTheme` environment. The effective dark/light is taken
    /// from `themeMode` (no separate neon mode setting).
    var neonPalette: NeonPaletteChoice {
        didSet { defaults.set(neonPalette.rawValue, forKey: Keys.neonPalette) }
    }

    /// Neon Terminal glow on/off. Persisted; flows into
    /// `NeonTheme.resolve(...)` so later card work can render (or skip)
    /// the layered glow shadows.
    var neonGlow: Bool {
        didSet { defaults.set(neonGlow, forKey: Keys.neonGlow) }
    }

    /// Base point size the typography ramp (`ConduitTypography`)
    /// scales off. Setter clamps into [bodyPointSizeRange] so an
    /// out-of-range value (corrupted defaults, future migration) can't
    /// blow out the layout. Persisted on every set.
    var bodyPointSize: CGFloat = AppearanceStore.defaultBodyPointSize {
        didSet {
            let clamped = bodyPointSize.clamped(to: Self.bodyPointSizeRange)
            if clamped != bodyPointSize {
                bodyPointSize = clamped
                return
            }
            defaults.set(Double(bodyPointSize), forKey: Keys.bodyPointSize)
        }
    }

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        // New default is serif (matches the Claude iOS chat reference);
        // existing installs that explicitly chose monospaced/system keep
        // their preference because the persisted rawValue still resolves.
        self.fontFamily = (defaults.string(forKey: Keys.font)
            .flatMap(FontFamily.init(rawValue:))) ?? .serif
        self.themeMode = (defaults.string(forKey: Keys.theme)
            .flatMap(ThemeMode.init(rawValue:))) ?? .system
        self.collapseTurns = defaults.object(forKey: Keys.collapseTurns) as? Bool ?? false
        self.experimentalNativeTerminal =
            defaults.object(forKey: Keys.experimentalNativeTerminal) as? Bool ?? false
        // Default flipped to `true` in the upstream-ui-cutover (this PR):
        // ConduitUI is now the production tree. The flag is kept around
        // (rather than being deleted entirely) so an emergency revert
        // is one line — flip the default back to `false` and ship a
        // hotfix. The legacy view tree itself is gone, so flipping the
        // flag without restoring `Sources/Views/` would just render a
        // blank screen; we'll delete the flag in the next PR once the
        // cutover has soaked.
        self.experimentalConduitUI =
            defaults.object(forKey: Keys.experimentalConduitUI) as? Bool ?? true
        let storedBody = defaults.object(forKey: Keys.bodyPointSize) as? Double
        self.bodyPointSize = CGFloat(storedBody ?? Double(Self.defaultBodyPointSize))
            .clamped(to: Self.bodyPointSizeRange)
        let storedGhosttySize = defaults.object(forKey: Keys.ghosttyFontSize) as? Double
        self.ghosttyFontSize = (storedGhosttySize ?? Self.defaultGhosttyFontSize)
            .clamped(to: Self.ghosttyFontSizeRange)
        self.ghosttyTerminalTheme = (defaults.string(forKey: Keys.ghosttyTerminalTheme)
            .flatMap(GhosttyTerminalTheme.init(rawValue:))) ?? .ghosttyDark
        self.neonPalette = (defaults.string(forKey: Keys.neonPalette)
            .flatMap(NeonPaletteChoice.init(rawValue:))) ?? .ice
        self.neonGlow = defaults.object(forKey: Keys.neonGlow) as? Bool ?? true
    }

    /// SwiftUI `.font` value to use for chat body text.
    func bodyFont() -> Font {
        switch fontFamily {
        case .serif:      return .system(.body, design: .serif)
        case .system:     return .system(.body)
        case .monospaced: return .system(.body, design: .monospaced)
        }
    }

    /// Force every active UIWindow to honour the current `themeMode`.
    /// Belt-and-suspenders alongside `.preferredColorScheme` — that
    /// modifier alone was flaky on runtime swaps (light↔dark and back-
    /// to-system would silently no-op when triggered from inside a
    /// sheet). Setting `overrideUserInterfaceStyle` on the window is the
    /// UIKit-native mechanism and propagates to every modally-presented
    /// VC, which is what Settings → Appearance needs.
    ///
    /// Hops to the main actor before touching UIKit. `themeMode.didSet`
    /// can fire from any context (e.g. a Swift Testing task pool that
    /// is not the main thread); without this hop, Main Thread Checker
    /// trips even when the test logic itself is fine, and the test
    /// process exits non-zero despite all assertions passing.
    func applyToWindows() {
        if Thread.isMainThread {
            MainActor.assumeIsolated { applyToWindowsOnMain() }
        } else {
            DispatchQueue.main.async { [weak self] in
                self?.applyToWindowsOnMain()
            }
        }
    }

    @MainActor
    private func applyToWindowsOnMain() {
        let style: UIUserInterfaceStyle
        switch themeMode {
        case .system: style = .unspecified
        case .light:  style = .light
        case .dark:   style = .dark
        }
        for scene in UIApplication.shared.connectedScenes {
            guard let windowScene = scene as? UIWindowScene else { continue }
            for window in windowScene.windows {
                window.overrideUserInterfaceStyle = style
            }
        }
    }
}

private extension Comparable {
    func clamped(to range: ClosedRange<Self>) -> Self {
        min(max(self, range.lowerBound), range.upperBound)
    }
}
